doc/programming/guidelines.rst

```diff
@@ -0,0 +1,291 @@
+(i.e., with Zeek: the units that you enter into your EVT file). The
+attribute can be omitted otherwise.
+
+.. rubric:: Turn unreferenced vectors into anonymous fields
+
+Implement vectors that don't need to be referenced from your code as
+:ref:`anonymous fields <anonymous_fields>` (i.e., do not provide a
+name for the field). Normal, named fields store the whole vector
+accumulating all parsed elements over the lifetime of the defining
+unit; whereas anonymous vector field forgo storage, because nobody
+could access the fields anyways.
+
+The most common pitfall here is top-level units that parse a sequence
+of PDUs:
+
+.. code::
+
+    public type MyPDUs = unit {
+        msgs: Message[]; # accumulates all messages until end of session
+        : Message[];     # anonymous field, no accumulation
+    };
+
+.. rubric::  Skip unused fields
+
+Don't parse data that is not needed. Parsing data into a field will
+cause a copy of that data into a dedicated memory location. Instead,
+use the :ref:`skip keyword <skip>` to discard data that isn't of
+interest, which leads the compiler to generate optimized code for many
+field types, including in particular ``bytes``, literals, and
+generally fields of fixed size. Examples:
+
+.. code::
+
+    public type Message = unit {
+        unused1: uint32;            # parse and store
+        unused2: skip uint32;       # skip over 4 bytes
+
+        unused3: bytes &eod;        # extract and store remaining data
+        unused4: skip bytes &eod;   # skip over remaining data
+    };
+
+.. rubric::  Avoid Spicy strings and string manipulation
+
+Use Spicy :ref:`strings <type_string>` sparsely in your analyzer, and
+stick to :ref:`bytes <type_bytes>` instead where you can. Typically,
+you would convert ``bytes`` to strings as late as possible just before
+presentation to the user (including passing to Zeek events). Strings
+are never needed for the parsing process itself.
+
+As a corollary, avoid Spicy string manipulation. Always
+manipulate/concatenate ``bytes`` and convert only the final result to
+Spicy strings, probably in a ``%done`` hook or an EVT event. In
+particular, format strings come with a significant overhead. In other
+words, avoid use of ``%s`` to generate strings from ``bytes``.
+
+.. rubric::  Don't use temporary variables of expensive types just for readability
+
+Don't use temporary variables just to improve readability. In
+particular, ``string`` and ``bytes`` (that are implemented as C++
+strings under the hood and can be expensive to use) need to be
+allocated and destroyed. This may introduce relevant overhead as it
+cannot be guaranteed that the C++ compiler will be able to optimize
+the code generated by Spicy. To improve readability, comments are the
+tool of choice. In theory, benchmarking the impact of using a
+temporary variable is a valid approach as well. However, the benefit
+of slightly increased readability most likely does not justify the
+effort of benchmarking.
```

Comment:
Especially the last sentence is weird since we say on top that benchmarking is the ultimate guideline so it is unclear to me why we talk about trade offs between readability and benchmarking at all. We could probably replace all this with: _If in doubt, benchmark._